staging/src/k8s.io/component-base/metrics/timing_histogram.go

```diff
@@ -0,0 +1,253 @@
+// but modifies the Help description prior to object instantiation.
+func (h *TimingHistogram) initializeDeprecatedMetric() {
+	h.TimingHistogramOpts.markDeprecated()
+	h.initializeMetric()
+}
+
+// WithContext allows the normal TimingHistogram metric to pass in context. The context is no-op now.
+func (h *TimingHistogram) WithContext(ctx context.Context) GaugeMetric {
+	return h.GaugeMetric
+}
+
+// timingHistogramVec is the internal representation of our wrapping struct around prometheus
+// TimingHistogramVecs.
+type timingHistogramVec struct {
+	*promext.TimingHistogramVec
+	*TimingHistogramOpts
+	nowFunc func() time.Time
+	lazyMetric
+	originalLabels []string
+}
+
+// NewTimingHistogramVec returns an object which satisfies kubeCollector and
+// wraps the promext.timingHistogramVec object.  Note well the way that
+// behavior depends on registration and whether this is hidden.
+func NewTimingHistogramVec(opts *TimingHistogramOpts, labels []string) PreContextAndRegisterableGaugeMetricVec {
+	return NewTestableTimingHistogramVec(time.Now, opts, labels)
+}
+
+// NewTestableTimingHistogramVec adds injection of the clock.
+func NewTestableTimingHistogramVec(nowFunc func() time.Time, opts *TimingHistogramOpts, labels []string) PreContextAndRegisterableGaugeMetricVec {
+	opts.StabilityLevel.setDefaults()
+
+	fqName := BuildFQName(opts.Namespace, opts.Subsystem, opts.Name)
+	allowListLock.RLock()
+	if allowList, ok := labelValueAllowLists[fqName]; ok {
+		opts.LabelValueAllowLists = allowList
+	}
+	allowListLock.RUnlock()
+
+	v := &timingHistogramVec{
+		TimingHistogramVec:  noopTimingHistogramVec,
+		TimingHistogramOpts: opts,
+		nowFunc:             nowFunc,
+		originalLabels:      labels,
+		lazyMetric:          lazyMetric{},
+	}
+	v.lazyInit(v, fqName)
+	return v
+}
+
+// DeprecatedVersion returns a pointer to the Version or nil
+func (v *timingHistogramVec) DeprecatedVersion() *semver.Version {
+	return parseSemver(v.TimingHistogramOpts.DeprecatedVersion)
+}
+
+func (v *timingHistogramVec) initializeMetric() {
+	v.TimingHistogramOpts.annotateStabilityLevel()
+	v.TimingHistogramVec = promext.NewTestableTimingHistogramVec(v.nowFunc, v.TimingHistogramOpts.toPromHistogramOpts(), v.originalLabels...)
+}
+
+func (v *timingHistogramVec) initializeDeprecatedMetric() {
+	v.TimingHistogramOpts.markDeprecated()
+	v.initializeMetric()
+}
+
+// WithLabelValuesChecked, if called on a hidden vector,
+// will return a noop gauge and a nil error.
+// If called before this vector has been registered in
+// at least one registry, will return a noop gauge and
+// an error that passes ErrIsNotRegistered.
+// If called with a syntactic problem in the labels, will
+// return a noop gauge and an error about the labels.
+// If none of the above apply, this method will return
+// the appropriate vector member and a nil error.
+func (v *timingHistogramVec) WithLabelValuesChecked(lvs ...string) (GaugeMetric, error) {
+	if v.IsHidden() {
```

Comment:
OK - thanks for the explanation in:
https://github.com/kubernetes/kubernetes/pull/109729#issuecomment-1121353342

That makes sense to me and I'm fine with that.
But I want @dgrisonnet or @logicalhan to chime in here too.

Reply:
The approach taken to noop the metrics until they are initialized is correct. The deferred initialization of the metrics is normal behavior since the Kubernetes version is held by the Regsitry and because of that, it is impossible to determine if a metric is deprecated until it is registered.

As per handling whether the metrics is Hidden or not, that shouldn't be necessary since this is already taken care of when creating the underlying metric.
- some preprocessing will determine if the metric should be hidden: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/metrics/metric.go#L90-L100
- after this preprocessing, if the metrics is marked as hidden, it will not be created: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/metrics/metric.go#L146-L149